earthfile2llb/converter.go

```diff
@@ -477,14 +476,25 @@ func (c *Converter) SaveImage(ctx context.Context, imageNames []string, pushImag
 		justCacheHint = true
 	}
 	for _, imageName := range imageNames {
+		var st llb.State
+		if pushImages && c.mts.Final.RunPush.Initialized() {
```

Comment:
Yep. Those are the 4 cases; I have them enumerated in my tests.

1 is normal. It just goes in the main phase.

I made 2 go in the main phase, because that more or less is what should happen - we don't execute the push phase in builder if the `--push` option is not specified; which means it would never get saved locally if I put it in the push phase. It was the best way to break the catch-22

3 and 4 use the state from RunPush; since that has the State generated from the `RUN --push` commands, otherwise those changes are not reflected in the image. If they are based on the Main; It wont have that additional state.

Reply:
Right, but 3 and 4 although they use the state of the push phase, they still go into the main phase SaveLocal slice. That seems incorrect?

Also, I think the if condition should be

```suggestion
		if c.mts.Final.RunPush.Initialized() {
```

in order to treat 3 and 4 the same way.

Reply:
The condition could be shortened.... I was trying to explicitly clear here to avoid hidden dependencies. I have no strong feelings.

Also #3 is invalid looking closer at it from VSCode. That would not be allowed; Listener would deny it since its not a `--push` command.

Reply:
Ah ok - yeah - maybe best if 3 is not allowed. Then that clears up my understanding on the difference between 3 and 4.

What about this part:

> Right, but 3 and 4 although they use the state of the push phase, they still go into the main phase SaveLocal slice. That seems incorrect?

Reply:
Im working on it

Reply:
Done